### PR TITLE
Added execution of LLVM-IR for integration testing

### LINKED ISSUE #99: Implement LLVM-IR test harness
# Description
We need to create a crate to test LLVM-IR code.

# Spec

- [x] Execute LLVM-IR code with `lli`, extract the return value and the final memory status.
- [ ] Inspect the final value of any memory cell.

### PR DESCRIPTION
**Pull request by Eagle941**
# Summary

This PR adds the functions to compile and execute LLVM-IR and Cairo code and assert that the exit codes are matching.

# Details

It is assumed the Cairo code returns a single `u64` return value because that's what LLVM-IR (and traditional programs) return.

A new package `bzip2` has been added to the Darwin Nix workspace as required by `cairo-vm`.
A new file `CREDITS.md` has been added to the repo for compliance with MPL-2.0 and BSD-3-Clause licenses.

This PR closes #99.

# Checklist

- [x] Code is formatted by Rustfmt or `scarb fmt`.
- [x] Documentation has been updated if necessary.

### CODE REVIEW

File: CREDITS.md [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+The following software libraries require a dedicated copyright notice for distribution.
+
+#### MPL-2.0
+
+- **MPL-2.0** &lt;<http://crates.io/crates/colored>&gt;
+- **MPL-2.0** &lt;<http://crates.io/crates/indent>&gt;
+
+#### BSD-3-Clause
+
+- **BSD-3-Clause** &lt;<http://crates.io/crates/subtle>&gt;
```
Comment (iamrecursion) [inline CREDITS.md:12]: This seems really strangely formatted. You have a section header for the license and then bold the license ahead of each library. This is redundant and you just need to mention the correct license once per library.

File: crates/test-utils/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,28 @@
+[package]
+name = "hieratika-test-utils"
+version.workspace = true
+homepage.workspace = true
+repository.workspace = true
+license-file.workspace = true
+
+authors.workspace = true
+description = "The utilities to test Hieratika."
```
Comment (iamrecursion) [inline crates/test-utils/Cargo.toml:9]:
```suggestion
description = "Utilities for E2E testing of the hieratika pipeline."
```

File: crates/test-utils/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# Make these specifically impossible to publish as they have no independent use.
+publish = false
+
+[dependencies]
+anyhow.workspace = true
+cairo-lang-compiler.workspace = true
+cairo-lang-runner = { path = "../../cairo/crates/cairo-lang-runner" }
+cairo-lang-sierra = { path = "../../cairo/crates/cairo-lang-sierra" }
+cairo-lang-sierra-generator = { path = "../../cairo/crates/cairo-lang-sierra-generator" }
+cairo-lang-utils = { path = "../../cairo/crates/cairo-lang-utils" }
```
Comment (iamrecursion) [inline crates/test-utils/Cargo.toml:25]: I feel like these deps may end up used in more than one place, so please hoist them to workspace level and use them here that way. Otherwise we just end up with people duplicating these paths across multiple Cargo projects which is ugly and harder to maintain if something changes.

File: crates/test-utils/README.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# Hieratika Test Utils
+
+Handy functions to simplify testing of Hiratika.
```
Comment (iamrecursion) [inline crates/test-utils/README.md:3]: This says nothing. It should explain what utilities it provides, primarily the E2E testing stuff.

File: crates/test-utils/input/square.cairo [hunk 5]
```diff
@@ -0,0 +1,7 @@
+fn square(a: i32) -> i32 {
+  a*a
+}
+
+fn main() -> i32 {
+  return square(3);
+}
```
Comment (iamrecursion) [inline crates/test-utils/input/square.cairo:7]: This should not live in the utils crate. The main E2E tests should probably live in "driver", and so should the test inputs.

File: crates/test-utils/input/square.ll [hunk 6]
```diff
@@ -0,0 +1,17 @@
+define dso_local i32 @square(i32) {
```
Comment (iamrecursion) [inline crates/test-utils/input/square.ll:1]: This should not live in the utils crate. The main E2E tests should probably live in "driver", and so should the test inputs.

Also, where on earth did this come from? It needs a comment saying it. 

It is also really strange LLVM IR that is not really equivalent to the above cairo.

File: crates/test-utils/input/square.ll [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  %2 = alloca i32, align 4
+  store i32 %0, i32* %2, align 4
+  %3 = load i32, i32* %2, align 4
+  %4 = load i32, i32* %2, align 4
+  %5 = mul nsw i32 %3, %4
+  ret i32 %5
+}
+
+declare void @llvm.dbg.declare(metadata, metadata, metadata) #1
```
Comment (iamrecursion) [inline crates/test-utils/input/square.ll:10]: This is not used here as far as I can see. The declaration is spurious.

File: crates/test-utils/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+    path: &Path,
+    function_name: &str,
+    args: &[&GenericValue],
+) -> anyhow::Result<u64> {
+    let context = Context::create();
+    let memory_buffer = MemoryBuffer::create_from_file(path).map_err(Error::from)?;
+    let module = context.create_module_from_ir(memory_buffer).map_err(Error::from)?;
+    let execution_engine = module
+        .create_jit_execution_engine(OptimizationLevel::None)
+        .map_err(Error::from)?;
```
Comment (iamrecursion) [inline crates/test-utils/src/lib.rs:49]: Does this create an LLI session, or a no-optimization ORC JIT? We specifically want LLI here, as it does not involve compilation to machine code for execution.

File: crates/test-utils/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+    let function = module
+        .get_function(function_name)
+        .unwrap_or_else(|| panic!("Function {function_name} not present in {path:?}"));
+
+    unsafe {
+        let output = execution_engine.run_function(function, args);
+        let output = output.as_int(false);
+        Ok(output)
+    }
```
Comment (iamrecursion) [inline crates/test-utils/src/lib.rs:58]: Is all of this unsafe? Certainly the `Ok(output)` does not need to be. We need to keep our `unsafe` blocks as small as possible.

File: crates/test-utils/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    values::GenericValue,
+    OptimizationLevel,
+};
+
+/// Executes the LLVM-IR program and returns the exit code as unsigned integer.
+///
+/// # Arguments
+///
+/// - `path` - The path of the LLVM-IR file to execute
+/// - `function_name` - The name of the function to execute
```
Comment (iamrecursion) [inline crates/test-utils/src/lib.rs:29]: Sentences end with a full stop.

File: crates/test-utils/src/lib.rs [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+        Ok(output)
+    }
+}
+
+/// Executes the Cairo program and returns the exit code as unsigned integer.
+///
+/// It is assumed the Cairo function returns a single `felt252` which can be
+/// casted to `u64` without overflow. The reason is that this function is used
+/// to verify that the equivalent LLVM-IR code returns the same exit code and
+/// standard exit codes are single integers.
```
Comment (iamrecursion) [inline crates/test-utils/src/lib.rs:66]: This is not an assumption that we should make of the Cairo code. The cairo code should be able to return a full `felt252`. Wanting to compare these for equality should be done _externally_. You could modify the `assert_eq_llvm_cairo` utility function for this if you wanted, but it is not the domain of a function for executing Cairo code.

File: crates/test-utils/src/lib.rs [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+        let SierraProgramWithDebug {
+            program: mut sierra_program,
+            debug_info: _debug_info,
+        } = Arc::unwrap_or_clone(
+            db.get_sierra_program(main_crate_ids).map_err(Error::SalsaDbError)?,
+        );
+        let replacer = DebugReplacer { db };
+        replacer.enrich_function_names(&mut sierra_program);
+        Ok(replacer.apply(&sierra_program))
+    }
```
Comment (iamrecursion) [inline crates/test-utils/src/lib.rs:184]: Why is this here? This should be exposed as a testing utility as well.

File: deny.toml [hunk 13]
```diff
@@ -101,6 +101,14 @@ exceptions = [
...
   { allow = ["Unicode-DFS-2016"], crate = "unicode-ident" },
   { allow = ["CC0-1.0"], crate = "tiny-keccak" },
+  { allow = ["MPL-2.0"], crate = "colored" },
+  { allow = ["Zlib"], crate = "foldhash" },
+  { allow = ["MPL-2.0"], crate = "indent" },
+  { allow = ["BSD-3-Clause"], crate = "subtle" },
+  { allow = ["MIT"], crate = "unescaper" },
+  { allow = ["Zlib"], crate = "const_format" },
+  { allow = ["Zlib"], crate = "const_format_proc_macros" },
+  { allow = ["CC0-1.0"], crate = "constant_time_eq" },
```
Comment (iamrecursion) [inline deny.toml:111]: These are more crates than you list in the `CREDITS.md` file. Are the ones not listed okay without any special action?

**Review by iamrecursion — CHANGES_REQUESTED** [review]
This is an excellent start, but there are a few things I want to see tweaked and clarified first!

File: deny.toml (hunk 13 above)
Comment (Eagle941) [reply deny.toml:111]: Yes only MPL-2 and BSD-3 libraries need to be listed for the time being.

File: crates/test-utils/src/lib.rs (hunk 8 above)
Comment (Eagle941) [reply crates/test-utils/src/lib.rs:49]: Ahhh ok! I thought it was best to use the JIT engine for interpreted execution.

File: crates/test-utils/input/square.ll (hunk 6 above)
Comment (Eagle941) [reply crates/test-utils/input/square.ll:1]: I have generated the LLVM-IR code from the Rust compiler using a program equivalent to the Cairo code.

Comment (iamrecursion) [reply crates/test-utils/input/square.ll]: This is not really equivalent. It is doing things that the cairo code does not with memory. I would suggest hand-writing it instead, and avoiding the pointers to just use SSA immediates instead. That would be more equivalent.

File: deny.toml (hunk 13 above)
Comment (iamrecursion) [reply deny.toml:111]: Okay.

**Review by iamrecursion — APPROVED** [review]
LGTM.
